src/components/MusicInputFormContainer.js

```diff
@@ -0,0 +1,347 @@
+import React, { Component } from "react";
+import { connect } from "react-redux";
+import ReactAbcjs from "./react-abcjs.js";
+import "./MusicInputForm.css";
+import { submitAnswer, loadStats } from "../actions/dictation";
+import MidiPlayer from "web-midi-player";
+import ABCJS from "abcjs/midi";
+
+class MusicInputFormContainer extends Component {
+  initialState = {
+    original: this.props.melody.abcStart,
+    userInput: [],
+    notes: []
+  };
+
+  state = this.initialState;
+
+  midiPlayer = new MidiPlayer();
+
+  pitchesAndRegex = [
+    ["F,", /F,/],
+    ["G,", /G,/],
+    ["A,", /A,/],
+    ["B,", /B,/],
+    ["C", /C(?!,)(?!')/],
+    ["D", /D(?!,)(?!')/],
+    ["E", /E(?!,)(?!')/],
+    ["F", /F(?!,)(?!')/],
+    ["G", /G(?!,)(?!')/],
+    ["A", /A(?!,)(?!')/],
+    ["B", /B(?!,)(?!')/],
+    ["c", /c(?!,)(?!')/],
+    ["d", /d(?!,)(?!')/],
+    ["e", /e(?!,)(?!')/],
+    ["f", /f(?!,)(?!')/],
+    ["g", /g(?!,)(?!')/],
+    ["a", /a(?!,)(?!')(?!r)/],
+    ["b", /b(?!,)(?!')/],
+    ["c'", /c'/],
+    ["d'", /d'/],
+    ["e'", /e'/]
+  ];
+
+  signs = [
+    ["^", "\ue262"],
+    ["_", "\ue260"],
+    ["=", "\ue261"]
+  ];
+
+  playAbc = async abcNotation => {
+    ABCJS.renderMidi(
+      "abc",
+      abcNotation,
+      {},
+      { generateInline: false, generateDownload: true },
+      {}
+    );
+    const myDiv = document.getElementById("abc");
+    const userInputMidiUrl = myDiv.children[0].children[0].getAttribute("href");
+    await this.midiPlayer.stop();
+    await this.midiPlayer.play({ url: userInputMidiUrl });
+  };
+
+  increasePitchOfLastNote = () => {
+    const lastNote = this.state.userInput[this.state.userInput.length - 1];
+
+    if (this.state.userInput.length > 0 && lastNote !== "|") {
+      this.pitchesAndRegex.forEach((pitch, index, array) => {
+        const notReachHighestLine = index < this.pitchesAndRegex.length - 1;
+        const pitchOfLastNote = lastNote.match(pitch[1]);
+        if (pitchOfLastNote && notReachHighestLine) {
+          this.setState({
+            ...this.state,
+            userInput: [
+              ...this.state.userInput.slice(0, -1),
+              lastNote.replace(
+                pitchOfLastNote,
+                this.pitchesAndRegex[index + 1][0]
+              )
+            ]
+          });
+        }
+      });
+    }
+  };
+  decreasePitchOfLastNote = () => {
+    const lastNote = this.state.userInput[this.state.userInput.length - 1];
+    if (this.state.userInput.length > 0 && lastNote !== "|") {
+      this.pitchesAndRegex.forEach((pitch, index, array) => {
+        const notReachLowestLine = index > 0;
+        const pitchOfLastNote = lastNote.match(pitch[1]);
+        if (pitchOfLastNote && notReachLowestLine) {
+          this.setState({
```

Comment:
`setState` inside loop looks bad. It much better calculates new data and passes it one time